### ISSUE TITLE
SmallVec can not coerce lifetimes of wrapped value

### ISSUE THREAD

**Issue by Luro02** [servo/rust-smallvec#217]
I could not get the following code to compile with `SmallVec` (I heavily simplified the code):

```rust
use std::borrow::Cow;

pub struct SmallVec<T> {
    data: *mut T,
}

pub struct Example<'a>(SmallVec<[Cow<'a, str>; 2]>);

impl<'a> Example<'a> {
    pub fn into_owned(self) -> Example<'static> { unimplemented!() }
}

fn transform<'a>(input: Example<'a>) -> Example<'a> {
    input.into_owned()
}

fn main() {}
```
[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&code=use%20std%3A%3Aborrow%3A%3ACow%3B%0A%0Apub%20struct%20SmallVec%3CT%3E%20%7B%0A%20%20%20%20data%3A%20*mut%20T%2C%0A%7D%0A%0Apub%20struct%20Example%3C'a%3E(SmallVec%3C%5BCow%3C'a%2C%20str%3E%3B%202%5D%3E)%3B%0A%0Aimpl%3C'a%3E%20Example%3C'a%3E%20%7B%0A%20%20%20%20pub%20fn%20into_owned(self)%20-%3E%20Example%3C'static%3E%20%7B%20unimplemented!()%20%7D%0A%7D%0A%0Afn%20transform%3C'a%3E(input%3A%20Example%3C'a%3E)%20-%3E%20Example%3C'a%3E%20%7B%0A%20%20%20%20input.into_owned()%0A%7D%0A%0Afn%20main()%20%7B%7D%0A)

```
error[E0308]: mismatched types
  --> src/main.rs:14:5
   |
14 |     input.into_owned()
   |     ^^^^^^^^^^^^^^^^^^ lifetime mismatch
   |
   = note: expected struct `Example<'a>`
              found struct `Example<'static>`
note: the lifetime `'a` as defined on the function body at 13:14...
  --> src/main.rs:13:14
   |
13 | fn transform<'a>(input: Example<'a>) -> Example<'a> {
   |              ^^
   = note: ...does not necessarily outlive the static lifetime

error: aborting due to previous error
```
which does compile when you replace `SmallVec` with `Vec`.

This works with `Vec`, because internally the pointer is wrapped in `Unique<T>`:

https://github.com/rust-lang/rust/blob/fb5615a4771ea3d54256f969dc84d2dfd38d812c/src/liballoc/raw_vec.rs#L47

whereas `SmallVec` contains a raw pointer:

https://github.com/servo/rust-smallvec/blob/a541e2d0dc39c31a9f0f9ed6675d26b79079a5aa/lib.rs#L370

This issue can be solved by changing `*mut A::Item` to `Option<NonNull<A::Item>>`, because `NonNull` implements the [`CoerceUnsized`](https://doc.rust-lang.org/stable/std/ptr/struct.NonNull.html#impl-CoerceUnsized%3CNonNull%3CU%3E%3E) trait.

This might be related to #146

**Comment by alejandro-vaz** [comment]
yeah we already moved to `NonNull` on v2, and v1 has its API locked down
